File: tradeit-android-sdk/src/main/java/it/trade/android/sdk/TradeItConfigurationBuilder.java
package it.trade.android.sdk;

import android.content.Context;

import it.trade.android.sdk.model.RequestInterceptorParcelable;
import it.trade.model.request.TradeItEnvironment;

public class TradeItConfigurationBuilder {

    private Context context;
    private String apiKey;
    private TradeItEnvironment environment;
    private String baseUrl;
    private RequestInterceptorParcelable requestInterceptorParcelable;

    private TradeItConfigurationBuilder() {
    }

    public TradeItConfigurationBuilder(Context context, String apiKey, TradeItEnvironment environment) {
        this.context = context;
        this.apiKey = apiKey;
        this.environment = environment;
    }

    public TradeItConfigurationBuilder withBaseUrl(String baseUrl) {
        this.baseUrl = baseUrl;
        return this;
    }

    public TradeItConfigurationBuilder withRequestInterceptor(RequestInterceptorParcelable requestInterceptorParcelable) {
        this.requestInterceptorParcelable = requestInterceptorParcelable;
        return this;
    }

    public Context getContext() {
        return context;
    }

    public String getApiKey() {
        return apiKey;
    }

    public TradeItEnvironment getEnvironment() {
        return environment;
    }

    public String getBaseUrl() {
        return baseUrl;
    }

    public RequestInterceptorParcelable getRequestInterceptorParcelable() {
        return requestInterceptorParcelable;
    }
}

File: tradeit-android-sdk/src/main/kotlin/it/trade/android/sdk/TradeItConfigurationBuilder.kt
package it.trade.android.sdk

import android.content.Context

import it.trade.android.sdk.model.RequestInterceptorParcelable
import it.trade.model.request.TradeItEnvironment

class TradeItConfigurationBuilder (
        val context: Context,
        val apiKey: String,
        val environment: TradeItEnvironment
    ) {

    var baseUrl: String? = null
        private set
    var requestInterceptorParcelable: RequestInterceptorParcelable? = null
        private set
    var isPrefetchBrokerListEnabled = true
        private set

    fun withBaseUrl(baseUrl: String): TradeItConfigurationBuilder {
        this.baseUrl = baseUrl
        return this
    }

    fun withRequestInterceptor(requestInterceptorParcelable: RequestInterceptorParcelable): TradeItConfigurationBuilder {
        this.requestInterceptorParcelable = requestInterceptorParcelable
        return this
    }

    fun disablePrefetchBrokerList(): TradeItConfigurationBuilder {
        this.isPrefetchBrokerListEnabled = false
        return this
    }
}
